src/pymmcore_widgets/mda/_core_positions.py

```diff
@@ -51,6 +52,54 @@ def __init__(
         toolbar.insertWidget(action0, self.move_to_selection)
         self.table().itemSelectionChanged.connect(self._on_selection_change)
 
+        # connect
+        self._mmc.events.systemConfigurationLoaded.connect(self._on_sys_config_loaded)
+        self._mmc.events.propertyChanged.connect(self._on_property_changed)
+
+        self.destroyed.connect(self._disconnect)
+
+        self._on_sys_config_loaded()
+
+    def _on_sys_config_loaded(self) -> None:
+        """Update the table when the system configuration is loaded."""
+        self._enable_xy()
+        self._enable_z()
+
+    def _enable_xy(self) -> None:
+        """Enable/disable the XY columns and button."""
+        xy_device = self._mmc.getXYStageDevice()
+        x_col = self.table().indexOf(self.X)
+        y_col = self.table().indexOf(self.Y)
+        self.table().setColumnHidden(x_col, not bool(xy_device))
+        self.table().setColumnHidden(y_col, not bool(xy_device))
+        xy_btn_col = self.table().indexOf(self._xy_btn_col)
+        self.table().setColumnHidden(xy_btn_col, not bool(xy_device))
+
+    def _enable_z(self) -> None:
+        """Enable/disable the Z columns and button."""
+        z_device = self._mmc.getFocusDevice()
+        self.include_z.setChecked(bool(z_device))
+        self.include_z.setEnabled(bool(z_device))
+        self.include_z.setToolTip("" if z_device else "No Focus device selected.")
+
+    def _on_property_changed(self, device: str, prop: str, value: str) -> None:
+        """Update the autofocus device combo box when the autofocus device changes."""
+        if device != "Core" or prop not in {"XYStage", "Focus"}:
+            return
+        if prop == "XYStage":
+            self._enable_xy()
+        elif prop == "Focus":
+            self._enable_z()
+
+    def _add_row(self) -> None:
```

Comment:
this is definitely super important, but it's tricky to implement.

As is currently, the valueChanged signal will be emitted before this update occurs, so anyone listening to valueChanged will get the wrong value for the positions.

Since the default/new values for a new row are provided by the ColumnInfo instance, I think it would be nice to have a `default_factory` option on ColumnInfo that can retrieve a default value as a function.  Then the values would be there *as* they get added to the table, and regardless of how the adding happened (button click or not).  Unfortunately, it would then be hard to pass the core instance to the default factory function.  So it might not be possible to use anything but the global core instance to populate the default values...

need to ponder this

Reply:
I just pushed some updates to this that should make sure this a) happens regardless of how the new row was added and b) only emits one signal per change